File: lib/object/objTo1d.ts
import type { AnyValueObject } from '../_types/global';
import { isObject } from '../verify/isObject.js';

export function objTo1d(obj: AnyValueObject, separator = '.'): AnyValueObject {
	if (!separator || separator.length < 1) {
		throw new Error('`separator` must have value at least 1 character.');
	}

	const convertObjectTo1d = (o: AnyValueObject, objPath = ''): AnyValueObject => {
		let result: AnyValueObject = {};
		const objectLength = Object.keys(o).length;
		const isFirstDepth = objPath.length < 1;

		for (let i = 0; i < objectLength; i += 1) {
			const key = Object.keys(o)[i];
			const value = o[key];
			const newObjPath = `${objPath}${isFirstDepth ? '' : separator}${key}`;

			if (isObject(value)) {
				result = Object.assign(result, convertObjectTo1d(value, newObjPath));
				delete result[key];
			} else {
				result[newObjPath] = value;
			}
		}

		return result;
	};

	return convertObjectTo1d(obj);
}

File: packages/python/qsu/object/objTo1d.py
from ..verify.isObject import isObject


def objTo1d(obj, separator='.'):
	if not separator or len(separator) < 1:
		raise Exception('`separator` must have value at least 1 character.')

	def convertObjectTo1d(o, objPath=''):
		result = {}
		isFirstDepth = len(objPath) < 1

		for key in o.keys():
			value = o[key]
			newObjPath = f"{objPath}{'' if isFirstDepth else separator}{key}"

			if isObject(value):
				result.update(convertObjectTo1d(value, newObjPath))
				if key in result:
					del result[key]
			else:
				result[newObjPath] = value

		return result

	return convertObjectTo1d(obj)
